storage/innobase/handler/i_s.cc

```diff
@@ -5055,8 +5055,11 @@ i_s_sys_tables_fill_table(
 		mutex_exit(&dict_sys.mutex);
 
 		if (!err_msg) {
-			i_s_dict_fill_sys_tables(thd, table_rec,
-						 tables->table);
+			if (int err = i_s_dict_fill_sys_tables(
+				thd, table_rec, tables->table)) {
+				ut_free(pcur.old_rec_buf);
+				DBUG_RETURN(err);
```

Comment:
This is missing `mem_heap_free(heap);`. Maybe, declare `int err` in a wider scope, move the release of `dict_sys.mutex` before `mtr_t::commit()` at the end of the function, and add `goto func_exit;` for the error handling.